File: leet-code/332. Reconstruct Itinerary/ReconstructItinerary332_2.java
package com.example;

import java.util.*;

public class ReconstructItinerary332_2 {
    public List<String> findItinerary(List<List<String>> tickets) {
        Map<String, List<String>> routes = new HashMap<>();
        for (List<String> ticket : tickets) {
            String from = ticket.get(0);
            String to = ticket.get(1);
            List<String> values = routes.getOrDefault(from, new ArrayList<>());
            values.add(to);
            routes.put(from, values);
        }
        for (List<String> value : routes.values()) {
            value.sort(String::compareTo);
        }

        List<String> answer = new ArrayList<>();
        dfs(routes, answer, "JFK");
        return answer;
    }

    private void dfs(Map<String, List<String>> routes,
                     List<String> history,
                     String departure) {
        while (routes.containsKey(departure) && !routes.get(departure).isEmpty()) {
            String arrival = routes.get(departure).remove(0);
            dfs(routes, history, arrival);
        }
        history.add(0, departure);
    }
}

File: leet-code/332. Reconstruct Itinerary/ReconstructItinerary332_2.kt
package com.example

import java.util.*

class ReconstructItinerary332_2 {
    fun findItinerary(tickets: List<List<String>>): List<String> {
        val map = mutableMapOf<String, PriorityQueue<String>>()
        for (ticket in tickets) {
            map.putIfAbsent(ticket[0], PriorityQueue<String>())
            map[ticket[0]]!!.add(ticket[1])
        }
        return traverse(map, mutableListOf(), "JFK")
    }

    private fun traverse(
        route: MutableMap<String, PriorityQueue<String>>,
        itinerary: MutableList<String>,
        departure: String
    ): List<String> {
        while (route.containsKey(departure) && route[departure]!!.isNotEmpty()) {
            traverse(route, itinerary, route[departure]!!.poll())
        }
        itinerary.add(0, departure)
        return itinerary
    }
}
